File: src/antares_cli/core/sensitive_paths.py
"""Repository-local sensitive path policy."""

from __future__ import annotations

from pathlib import Path, PurePosixPath

_SENSITIVE_DIRECTORY_NAMES = frozenset(
    {
        ".aws",
        ".azure",
        ".docker",
        ".gnupg",
        ".kube",
        ".ssh",
        "gcloud",
    }
)
_SENSITIVE_FILE_NAMES = frozenset(
    {
        ".git-credentials",
        ".netrc",
        ".npmrc",
        ".pypirc",
        "application_default_credentials.json",
        "credentials",
        "id_dsa",
        "id_ecdsa",
        "id_ed25519",
        "id_rsa",
        "key.json",
        "kubeconfig",
    }
)
_SENSITIVE_FILE_SUFFIXES = (".jks", ".key", ".keystore", ".p12", ".pem", ".pfx")


def is_sensitive_repository_path(relative_path: str | PurePosixPath) -> bool:
    """Return whether a repository-relative path is protected by default."""
    normalized = str(relative_path).replace("\\", "/").casefold()
    while normalized.startswith("./"):
        normalized = normalized[2:]
    path = PurePosixPath(normalized)
    name = path.name
    parts = path.parts
    if name.startswith(".env"):
        return True
    if any(part in _SENSITIVE_DIRECTORY_NAMES for part in parts):
        return True
    if name in _SENSITIVE_FILE_NAMES or name.endswith(_SENSITIVE_FILE_SUFFIXES):
        return True
    return name.endswith(".json") and (
        name.startswith("service-account") or name.endswith("-credentials.json")
    )


def resolve_allowed_sensitive_files(
    source_root: str | Path,
    requested_paths: list[str] | tuple[str, ...] = (),
) -> tuple[str, ...]:
    """Validate exact runtime opt-ins and return canonical repository-relative paths."""
    root = Path(source_root).resolve()
    resolved_paths: list[str] = []
    for requested_path in requested_paths:
        raw_path = str(requested_path).strip()
        if not raw_path:
            raise ValueError("--allow-sensitive-file requires a non-empty relative file path")
        path = Path(raw_path)
        if path.is_absolute():
            raise ValueError(f"Sensitive-file opt-in must be repository-relative: {raw_path}")
        if ".." in PurePosixPath(raw_path.replace("\\", "/")).parts:
            raise ValueError(f"Sensitive-file opt-in does not allow parent traversal: {raw_path}")
        if any(character in raw_path for character in "*?[]"):
            raise ValueError(f"Sensitive-file opt-in does not allow glob patterns: {raw_path}")

        candidate = root / path
        try:
            relative_parts = candidate.relative_to(root).parts
        except ValueError as error:
            raise ValueError(
                f"Sensitive-file opt-in is outside the repository: {raw_path}"
            ) from error
        current = root
        for part in relative_parts:
            current /= part
            if current.is_symlink():
                raise ValueError(f"Sensitive-file opt-in does not allow symlinks: {raw_path}")
        try:
            resolved_candidate = candidate.resolve(strict=True)
            relative_path = resolved_candidate.relative_to(root).as_posix()
        except (FileNotFoundError, ValueError) as error:
            raise ValueError(
                f"Sensitive-file opt-in must name an existing file inside the repository: {raw_path}"
            ) from error
        if not resolved_candidate.is_file():
            raise ValueError(f"Sensitive-file opt-in must name a regular file: {raw_path}")
        if not is_sensitive_repository_path(relative_path):
            raise ValueError(
                f"Sensitive-file opt-in is unnecessary for an unprotected path: {relative_path}"
            )
        if relative_path not in resolved_paths:
            resolved_paths.append(relative_path)
    return tuple(resolved_paths)

File: vscode-extension/src/antares/core/sensitivePaths.ts
// Port of antares_cli/core/sensitive_paths.py — repository-local sensitive path policy.

const SENSITIVE_DIRECTORY_NAMES = new Set([
  ".aws",
  ".azure",
  ".docker",
  ".gnupg",
  ".kube",
  ".ssh",
  "gcloud",
]);

const SENSITIVE_FILE_NAMES = new Set([
  ".git-credentials",
  ".netrc",
  ".npmrc",
  ".pypirc",
  "application_default_credentials.json",
  "credentials",
  "id_dsa",
  "id_ecdsa",
  "id_ed25519",
  "id_rsa",
  "key.json",
  "kubeconfig",
]);

const SENSITIVE_FILE_SUFFIXES = [".jks", ".key", ".keystore", ".p12", ".pem", ".pfx"];

// Return whether a repository-relative path is protected by default.
export function isSensitiveRepositoryPath(relativePath: string): boolean {
  let normalized = relativePath.replace(/\\/g, "/").toLowerCase();
  while (normalized.startsWith("./")) {
    normalized = normalized.slice(2);
  }
  const parts = normalized.split("/").filter((part) => part.length > 0);
  const name = parts.length > 0 ? parts[parts.length - 1] : "";
  if (name.startsWith(".env")) {
    return true;
  }
  if (parts.some((part) => SENSITIVE_DIRECTORY_NAMES.has(part))) {
    return true;
  }
  if (SENSITIVE_FILE_NAMES.has(name) || SENSITIVE_FILE_SUFFIXES.some((s) => name.endsWith(s))) {
    return true;
  }
  return (
    name.endsWith(".json") &&
    (name.startsWith("service-account") || name.endsWith("-credentials.json"))
  );
}
